internal/controller/cert_controller.go

```diff
@@ -0,0 +1,276 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kubewarden/kubewarden-controller/internal/certs"
+	"github.com/kubewarden/kubewarden-controller/internal/constants"
+)
+
+const tickerDuration = 12 * time.Hour
+
+type CertReconciler struct {
+	client.Client
+	Log                         logr.Logger
+	DeploymentsNamespace        string
+	WebhookServiceName          string
+	CARootSecretName            string
+	WebhookServerCertSecretName string
+}
+
+// Start begins the periodic reconciler.
+// Implements the Runnable inteface, see https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager#Runnable.
+func (r *CertReconciler) Start(ctx context.Context) error {
+	r.Log.Info("Starting CertController ticker")
+
+	ticker := time.NewTicker(tickerDuration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			r.Log.Info("Stopping CertController")
+			return nil
+		case <-ticker.C:
+			if err := r.reconcile(ctx); err != nil {
+				r.Log.Error(err, "Failed to")
+			}
+		}
+	}
+}
+
+// NeedLeaderElection returns true to ensure that only one instance of the controller is running at a time.
+// Implements the LeaderElectionRunnable interface, see https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager#LeaderElectionRunnable.
+func (r *CertReconciler) NeedLeaderElection() bool {
+	return true
+}
+
+func (r *CertReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := mgr.Add(r); err != nil {
+		return fmt.Errorf("failed enrolling controller with manager: %w", err)
+	}
+
+	return nil
+}
+
+// reconcile reconciles the CA root and server certificates by rotating them if they are about to expire.
+func (r *CertReconciler) reconcile(ctx context.Context) error {
+	caCertSecret := &corev1.Secret{}
+	if err := r.Get(ctx, types.NamespacedName{Name: r.CARootSecretName, Namespace: r.DeploymentsNamespace}, caCertSecret); err != nil {
+		return fmt.Errorf("failed to get CA cert secret: %w", err)
+	}
+
+	if err := r.reconcileCARoot(ctx, caCertSecret); err != nil {
+		return fmt.Errorf("failed to reconcile CA root: %w", err)
+	}
+	if err := r.reconcileOldCARoot(ctx, caCertSecret); err != nil {
+		return fmt.Errorf("failed to reconcile old CA root: %w", err)
+	}
+	if err := r.reconcileServerCerts(ctx, caCertSecret); err != nil {
```

Comment:
I would like to add a comment here to remind that everything hot reloads the certs. Something like:

> both policy-server and the controller webhook server certs support hot-reloading of certs. Hence, at this point, if we have removed the old cert after its expiration, we can be sure that policy-servers and the webhook server will reload the CA cert with the updated chain.